Add Reparameterization Checks

rbharath opened the pull request:
Working towards #129. More details once ready.

rbharath commented on tests/creatorblocktest/scenariosix/deploy_scenario_six_test.go:
```diff
@@ -55,7 +54,7 @@ func TestMain(m *testing.M) {
 	context.Blockchain = backends.NewSimulatedBackend(context.Alloc, 4700000)
 
-	deployed, deployedError = test.Deploy(zeroEth,
+	deployed, deployedError = test.Deploy(big.NewInt(test.ONE_ETH),
```
This test was failing because we were deploying with 0 market token supply, causing the desired stake to fail the correctness check. Tweaking so starting with nonzero markettoken supply.

rbharath commented:
This PR is now ready for review. It adds guard rails around parameterization in the constructor, reparamterization proposal, and reparameterization resolution. It should now be impossible for a market to ever have invalid parameters.

The PR fixes a broken test that relied on an incorrect parameterization, and adds a number of tests to check that invalid reparameterizations are not allowed.

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -45,17 +51,28 @@ backend_payment: uint256
 @public
-def __init__(v_addr: address, pr_fl: wei_value, spd: uint256, list_re: wei_value, stk: wei_value,
-  vote_by_d: timedelta, pl: uint256, back_p: uint256, maker_p: uint256, cost: wei_value):
+def __init__(v_addr: address, mkt_addr: address, pr_fl: wei_value, spd:
```
oh you know i've gone to _painful_ ends to make sure that _every_ operation that contains multiple contracts or addresses goes _in deployment order_.

voting then market?

nope

order of args here should be `mkt_addr` _then_ `v_addr`

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -45,17 +51,28 @@ backend_payment: uint256
+    assert stk <= (self.market_token.totalSupply()/3)
     self.stake = stk
+    # There are 604800 seconds in a week, 1209600 in 2 weeks
+    assert vote_by_d <= 1209600 
```
two things:
1. check in with TRON (that's reid btw) on what this should be. just making a call that a market can _never_ set a vote to less than 2 weeks is some presumptuous shit (and seems overlong to me...)
2. save it in a CONST -> SECONDS_IN_A_WEEK (or something)

rbharath replied:
This makes the opposite call that a market can never set a vote to _longer_ than 2 weeks. Will check with reid offline :)

Good call on the constant.

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -45,17 +51,28 @@ backend_payment: uint256
     self.spread = spd
     self.list_reward = list_re
+    # The stake can't exceed 1/3 of MarketToken supply
+    assert stk <= (self.market_token.totalSupply()/3)
```
let's abstract this one to a method -> we are doing it 3 times in this contract and its a little more involved than the other simple maths. obviously, replace the other instances with a call to the method...

`def get_max_stake` or something, may prove useful elsewhere as well...

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -186,6 +203,22 @@ def reparameterize(param: uint256, value: uint256):
+    # The spread is a percentage >= 100%
+    assert value >= 100 
+  elif param == STAKE:
+    assert value <= (self.market_token.totalSupply()/3)
```
`get_max_stake`

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -205,21 +238,28 @@ def resolveReparam(hash: bytes32):
       self.list_reward = value
     elif param == STAKE:
+      # Recheck validity in case MarketToken supply changed
+      assert value <= (self.market_token.totalSupply()/3)
```
`get_max_stake`

robrobbins commented on tests/test/deployed.go:
```diff
@@ -100,6 +100,7 @@ func Deploy(mtBal *big.Int, c *Ctx, p *Params) (*Dep, error) {
 		c.AuthOwner,
 		c.Blockchain,
 		votingAddr,
+		marketTokenAddr,
```
fix this order...

rbharath replied:
Order is fixed.

rbharath commented:
After offline review by @robrobbins, removed correctness checks from `Parameterizer.reparameterize` and placed them all in `Parameterizer.resolveReparam`. This simplifies the changes to the smart contracts, but comes with a couple costs:

- It's possible to spam nonsense reparams. This is fine for security since the resolution will catch them
- Adds some more boilerplate to tests

rbharath commented on contracts/Parameterizer.vy:
```diff
@@ -45,21 +56,45 @@ backend_payment: uint256
     self.list_reward = list_re
+    # The stake can't exceed 1/3 of MarketToken supply
+    # (We can't call getMaxStake() here since the function isn't declared
+    # yet)
```
Ended up getting rid of `getMaxStake()` since it was only used in one place now. Cleaner without it

rbharath commented on contracts/Parameterizer.vy:
```diff
@@ -45,21 +56,45 @@ backend_payment: uint256
 maker_payment: uint256
 cost_per_byte: wei_value
 voting: Voting
+market_token: MarketToken 
```
Done

rbharath commented:
@robrobbins I've removed the constructor checks and cleaned up the ordering issues (moving market token declarations above voting). I think this is ready for review again

robrobbins commented on contracts/Parameterizer.vy:
```diff
@@ -208,18 +220,29 @@ def resolveReparam(hash: bytes32):
       self.stake = value
     elif param == VOTE_BY:
+      assert value >= SECONDS_IN_DAY
+      assert value <= 2 * SECONDS_IN_WEEK 
```
noted above

    2 * (7 * SECONDS_IN_A_DAY)

rbharath replied:
Done

rbharath commented on contracts/Parameterizer.vy:
```diff
@@ -17,10 +17,20 @@ COST_PER_BYTE: constant(uint256) = 11
+# The number of seconds in a day
+SECONDS_IN_DAY: constant(uint256) = 86400 
+# The number of seconds in a week.
+SECONDS_IN_WEEK: constant(uint256) = 604800
```
Removed the week constant
